ishop/views.py

```diff
@@ -1,3 +1,179 @@
+        good.in_stock -= quantity
+        good.save()
+        Purchase.objects.create(customer=user, good=good, quantity=quantity, price=price)
+
+        msg = "Your purchase is done"
+        messages.success(self.request, msg)
+
+        return self.get(request, *args, **kwargs)
+
+
+class Account(LoginRequiredMixin, ListView):
+    model = Purchase
+    paginate_by = 10
+    template_name = 'account.html'
+    extra_context = {'title': 'My account'}
+    time_interval_to_refund = 3  # in minutes
+
+    def get_queryset(self):
+        queryset = self.model.objects.filter(customer=self.request.user)
+        return queryset
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        delta = timezone.now() - timezone.timedelta(minutes=self.time_interval_to_refund)
+        context['refund_possible'] = self.model.objects.filter(customer=self.request.user, datetime__gt=delta)
+        in_refund = Refund.objects.filter(purchase__in=self.get_queryset())
+        context['in_refund'] = in_refund.values_list('purchase__pk', flat=True)
+        context['balance'] = ShopUser.objects.get(id=self.request.user.id).wallet
+        return context
+
+    def post(self, request, *args, **kwargs):
+        purchase = Purchase.objects.get(pk=self.request.POST['pk'])
+        delta = timezone.now() - timezone.timedelta(minutes=self.time_interval_to_refund)
+        if purchase.datetime < delta:
+            msg = 'Your refund time has been expired'
+            messages.error(self.request, msg)
+            return self.get(request, *args, **kwargs)
+
+        _, is_new = Refund.objects.get_or_create(purchase=purchase)
+        if is_new:
+            msg = 'Your refund request has been sent. Wait for approving.'
+            messages.success(self.request, msg)
+        else:
+            msg = 'The refund for current purchase already exists. Wait for approving.'
+            messages.error(self.request, msg)
+
+        return self.get(request, *args, **kwargs)
+
+
+class AdminRefundView(SuperUserRequiredMixin, ListView):
+    model = Refund
+    paginate_by = 10
+    template_name = 'admin_refunds.html'
+    extra_context = {'title': 'Admin: Purchases to refund'}
+
+    def post(self, request, *args, **kwargs):
+        pk = self.request.POST['pk']
+        approval = self.request.POST['approval']
+
+        if approval == 'decline':
+            self.model.objects.get(pk=pk).delete()
+        else:
+            refund = self.model.objects.get(pk=pk)
+            user = refund.purchase.customer
+            good = refund.purchase.good
+            quantity = refund.purchase.quantity
+            price = refund.purchase.price
+
+            user.wallet += price * quantity
+            user.save()
+
+            good.in_stock += quantity
+            good.save()
+
+            self.model.objects.get(pk=pk).delete()
+            refund.purchase.delete()
+
+        return self.get(request, *args, **kwargs)
```

Comment:
return get for post action not a good idea (I know in framework we can find this too, but redirect is correct answer if you  aren't try to improve performance, in this case this discussible

Reply:
honestly, before adding my code, I copied the code of post-function from documentation and didn't pay attention to what it returns...
redirect would be better?

Reply:
changed